### PR TITLE
Add DDP Communication Hooks

### PR DESCRIPTION
**Pull request by yhna940**
## What does this PR do?

This PR adds support for DDP communication hooks to the `accelerate` library. Similar to frameworks like PyTorch Lightning and Detectron, these hooks provide an interface to control how gradients are communicated across workers, overriding the standard allreduce in DistributedDataParallel. This feature enables the use of performance-improving communication hooks when using multiple nodes.

### Motivation and Context

DDP communication hooks allow users to customize and optimize gradient communication, potentially improving training performance in distributed settings.

Based on the official PyTorch documentation [here](https://pytorch.org/docs/stable/ddp_comm_hooks.html), I've implemented three default hooks: PowerSGD, FP16, and BF16. These hooks provide performance improvements in distributed training scenarios.

The implementation for registering these hooks was inspired by the PyTorch Lightning implementation, which can be found [here](https://github.com/Lightning-AI/pytorch-lightning/blob/06ea3a05716a6d1f4a96cfb25021accdd18d8146/src/lightning/pytorch/overrides/distributed.py#L61).

## Fixes # (issue)

N/A

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case: [here](https://discuss.huggingface.co/t/feature-request-add-ddp-communication-hooks/91142/1)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: src/accelerate/utils/dataclasses.py [hunk 1]
```diff
@@ -79,6 +102,27 @@ class AutocastKwargs(KwargsHandler):
...
+    Values:
+
+        - **NO** -- no communication hook
+        - **FP16** -- DDP communication hook to compress the gradients in FP16
+        - **BF16** -- DDP communication hook to compress the gradients in BF16
+        - **POWER_SGD** -- DDP communication hook to use PowerSGD
+        - **BATCHED_POWER_SGD** -- DDP communication hook to use batched PowerSGD
+    """
+
+    # Subclassing str as well as Enum allows the `DistributedType` to be JSON-serializable out of the box.
```
Comment (muellerzr) [inline src/accelerate/utils/dataclasses.py:118]: Copy/paste leftover :)

File: src/accelerate/utils/dataclasses.py [hunk 2]
```diff
@@ -79,6 +102,27 @@ class AutocastKwargs(KwargsHandler):
     cache_enabled: bool = None
 
 
+class DDPCommunicationHookType(str, BaseEnum):
```
Comment (muellerzr) [inline src/accelerate/utils/dataclasses.py:105]: Is there a reason this isn't just `BaseEnum`? (Can you check that works)

**Review by muellerzr — COMMENTED** [review]
Thanks! Overall this looks great to me, a few requests though:

1. Can we add an example in `examples/by_feature` specifically showcasing this usage
2. Can we add some documentation to the official docs on it. Ideally a `Usage Guide` if nothing else (though a `Concept Guide` too would be ideal!)

File: src/accelerate/utils/dataclasses.py (hunk 2 above)
Comment (yhna940) [reply src/accelerate/utils/dataclasses.py]: Thank you for the suggestion! I have tested it, and it works correctly using just `BaseEnum`. I will update the code accordingly.

File: examples/by_feature/ddp_comm_hook.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        "between fp16 and bf16 (bfloat16). Bf16 requires PyTorch >= 1.10."
+        "and an Nvidia Ampere GPU.",
+    )
+    # New Code #
+    parser.add_argument(
+        "--ddp_comm_hook",
+        type=str,
+        default="no",
+        choices=["no", "fp16", "bf16", "power_sgd", "batched_power_sgd"],
+        help="DDP Communication hook to use. Choose between no, fp16, bf16, power_sgd, and batched_power_sgd.",
```
Comment (muellerzr) [inline examples/by_feature/ddp_comm_hook.py:214]:
```suggestion
        help="DDP Communication hook to use. Choose between `no`, `fp16`, `bf16`, `power_sgd`, and `batched_power_sgd`.",
```

File: examples/by_feature/ddp_comm_hook.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    )
+    # New Code #
+    parser.add_argument(
+        "--ddp_comm_wrapper",
+        type=str,
+        default="no",
+        choices=["no", "fp16", "bf16"],
+        help="DDP Communication wrapper to use. Choose between no, fp16, and bf16.",
```
Comment (muellerzr) [inline examples/by_feature/ddp_comm_hook.py:222]:
```suggestion
        help="DDP Communication wrapper to use. Choose between `no`, `fp16`, and `bf16`.",
```

**Review by muellerzr — APPROVED** [review]
Thanks for doing this! cc @SunMarc for a second eye, and @stevhliu for docs :)

**Review by SunMarc — APPROVED** [review]
Thanks for the clean PR @yhna940  ! LGTM ! Just a nit, could you add some details about `comm_wrapper` or `comm_state_option` ?

**Comment by yhna940** [comment]
Thank you for the feedback and the suggestion @SunMarc. I have added more details about `comm_wrapper` and `comm_state_option` in the documentation. As you mentioned, `comm_state_option` is currently only applied to `PowerSGD` in this PR.

However, there are other state-using options such as `post_localSGD_hook` available in PyTorch. These hooks require specific optimizers and provide advanced features for gradient communication. For example, `post_localSGD_hook` is closely tied to the `PostLocalSGDOptimizer` and involves additional setup:

```python
import torch
import torch.distributed as dist
import torch.distributed.algorithms.model_averaging.averagers as averagers
import torch.nn as nn
from torch.distributed.optim import PostLocalSGDOptimizer
from torch.distributed.algorithms.ddp_comm_hooks.post_localSGD_hook import (
    PostLocalSGDState,
    post_localSGD_hook,
)

model = nn.parallel.DistributedDataParallel(
    module, device_ids=[rank], output_device=rank
)

# Register a post-localSGD communication hook.
state = PostLocalSGDState(process_group=None, subgroup=None, start_localSGD_iter=100)
model.register_comm_hook(state, post_localSGD_hook)

# Create a post-localSGD optimizer that wraps a local optimizer.
# Note that `warmup_steps` used in `PostLocalSGDOptimizer` must be the same as
# `start_localSGD_iter` used in `PostLocalSGDState`.
local_optim = torch.optim.SGD(params=model.parameters(), lr=0.01)
opt = PostLocalSGDOptimizer(
    optim=local_optim,
    averager=averagers.PeriodicModelAverager(period=4, warmup_steps=100)
)

# In the first 100 steps, DDP runs global gradient averaging at every step.
# After 100 steps, DDP runs gradient averaging within each subgroup (intra-node by default),
# and post-localSGD optimizer runs global model averaging every 4 steps after applying the local optimizer.
for step in range(0, 200):
    opt.zero_grad()
    loss = loss_fn(output, labels)
    loss.backward()
    opt.step()
```

These advanced hooks were not included in this PR because the PyTorch official documentation primarily highlights PowerSGD, FP16, and BF16 hooks. You can find more information about additional hooks in the [PyTorch DDP Communication Hooks documentation](https://pytorch.org/docs/stable/ddp_comm_hooks.html) and the [PyTorch GitHub repository](https://github.com/pytorch/pytorch/tree/main/torch/distributed/algorithms/ddp_comm_hooks).

**Comment by yhna940** [comment]
Thanks for the quick review @SunMarc , I've done it :)

File: docs/source/_toctree.yml [hunk 5]
```diff
@@ -58,6 +58,8 @@
       title: Apple M1 GPUs
     - local: usage_guides/ipex
       title: IPEX training with CPU
+    - local: usage_guides/ddp_comm_hook
```
Comment (stevhliu) [inline docs/source/_toctree.yml:61]: I would consider putting it after or before Fully Sharded Data Parallelism since they're more closely related.

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+
+⚠️ Note that this file is in Markdown but contains specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+-->
+
+# Performing DDP Communication Hooks with 🤗 Accelerate
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:17]: Align the title with the one in the toctree to avoid confusion when referencing the doc

```suggestion
# DDP Communication Hooks
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+DDP communication hooks provide a generic interface to control how gradients are communicated across workers by overriding the vanilla allreduce in `DistributedDataParallel`. A few built-in communication hooks are provided, and users can easily apply any of these hooks to optimize communication.
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:19]:
```suggestion
Distributed Data Parallel (DDP) communication hooks provide a generic interface to control how gradients are communicated across workers by overriding the vanilla allreduce in `DistributedDataParallel`. A few built-in communication hooks are provided, and users can easily apply any of these hooks to optimize communication.
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+## Brief Descriptions of DDP Communication Hooks
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:21]: The descriptions below can be a part of the overall "overview", rather than dividing the doc so granularly

```suggestion
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+- **FP16 Compression Hook**: Compresses gradients by casting them to half-precision floating-point format (`torch.float16`), reducing communication overhead.
+- **BF16 Compression Hook**: Similar to FP16, but uses the Brain Floating Point format (`torch.bfloat16`), which can be more efficient on certain hardware.
+- **PowerSGD Hook**: An advanced gradient compression algorithm that provides high compression rates and can accelerate bandwidth-bound distributed training.
+
+In this tutorial, you will see how to quickly set up DDP communication hooks and perform training with the utilities provided in 🤗 Accelerate, which can be as simple as adding just one new line of code!
+
+## Using DDP Communication Hooks with PyTorch's `DistributedDataParallel`
+
+### Using FP16 Compression Hook with PyTorch
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:31]: Probably not necessary to have the second level header (`##`) since it has no content directly after it. You could probably also shorten the third level header (`###`) a bit because users probably already understand this is for PyTorch.

```suggestion
## FP16 Compression Hook
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+# Training loop
+for data, targets in data_loader:
+    outputs = model(data)
+    loss = criterion(outputs, targets)
+    loss.backward()
+    optimizer.step()
+    optimizer.zero_grad()
+```
+
+### Using BF16 Compression Hook with PyTorch
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:59]:
```suggestion
### BF16 Compression Hook
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+# Training loop
+for data, targets in data_loader:
+    outputs = model(data)
+    loss = criterion(outputs, targets)
+    loss.backward()
+    optimizer.step()
+    optimizer.zero_grad()
+```
+
+### Using PowerSGD Hook with PyTorch
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:87]:
```suggestion
### PowerSGD Hook
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+    loss.backward()
+    optimizer.step()
+    optimizer.zero_grad()
+```
+
+## Converting it to 🤗 Accelerate
+
+Now, let's see how to use the same hooks with the 🤗 Accelerate library.
+
+### Using FP16 Compression Hook with 🤗 Accelerate
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:120]: You could use the `<hfoption>` tags to create separate tabs the user can click on to view the native version vs using it in Accelerate. It'll look like the tabs in this section [here](https://huggingface.co/docs/diffusers/tutorials/autopipeline#autopipeline). For example, for FP16 compression you can do:

```
<hfoptions id="fp16">
<hfoption id="PyTorch">

code

</hfoption>
<hfoption id="Accelerate">

code

</hfoption>
</hfoptions>
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+# Training loop
+for data, targets in data_loader:
+    outputs = model(data)
+    loss = criterion(outputs, targets)
+    accelerator.backward(loss)
+    optimizer.step()
+    optimizer.zero_grad()
+```
+
+## Additional Options for DDP Communication Hooks
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:231]: It is generally best to not have an empty section header :)

```suggestion
## DDP Communication Hooks utilities

There are two additional utilities for supporting optional functionalities with the communication hooks.
```

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+    outputs = model(data)
+    loss = criterion(outputs, targets)
+    accelerator.backward(loss)
+    optimizer.step()
+    optimizer.zero_grad()
+```
+
+For more advanced usage and additional hooks, refer to the [PyTorch DDP Communication Hooks documentation](https://pytorch.org/docs/stable/ddp_comm_hooks.html).
+
+This demonstrates how to use DDP communication hooks to optimize gradient communication in distributed training with the 🤗 Accelerate library.
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:277]: This sentence would be more impactful in the beginning where you describe what the tutorial will do.

File: docs/source/usage_guides/ddp_comm_hook.md [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+    accelerator.backward(loss)
+    optimizer.step()
+    optimizer.zero_grad()
+```
+
+## Additional Options for DDP Communication Hooks
+
+### comm_wrapper
+
+`comm_wrapper` is an option to wrap a communication hook with additional functionality. For example, it can be used to combine FP16 compression with other communication strategies. Currently supported wrappers are `no`, `fp16`, and `bf16`.
```
Comment (stevhliu) [inline docs/source/usage_guides/ddp_comm_hook.md:235]: This would be awesome if you can provide a code example!

**Review by stevhliu — APPROVED** [review]
Very nice job! 👏

**Comment by yhna940** [comment]
Thank you for your detailed review @stevhliu :) I've changed the guide as you suggested, I used `hfoption` option tag and rearranged some sentences, I'll fix it if there's anything wrong, thanks 🤗

**Comment by muellerzr** [comment]
Nicely done @yhna940!
